File: src/chatguide/utils/response_parser.py
"""Response parser - parses LLM responses."""

import json
from typing import Any
from ..schemas import ChatGuideReply


def parse_llm_response(raw: Any) -> ChatGuideReply:
    """Parse LLM response into ChatGuideReply."""
    if raw is None:
        raise ValueError("LLM returned no content")

    if isinstance(raw, ChatGuideReply):
        return raw
    
    if isinstance(raw, dict):
        return ChatGuideReply.model_validate(raw)
    
    if isinstance(raw, str):
        return ChatGuideReply.model_validate(json.loads(raw))
    
    raise ValueError(f"Unexpected response type: {type(raw)}")

File: js/src/utils/responseParser.ts
import { ChatGuideReply, ChatGuideReplySchema } from '../schemas';

export function parseLLMResponse(raw: any): ChatGuideReply {
  if (!raw) {
    throw new Error('LLM returned no content');
  }

  if (typeof raw === 'string') {
    try {
      raw = JSON.parse(raw);
    } catch {
      throw new Error('Invalid JSON response from LLM');
    }
  }

  // Handle task_id field from LLM responses and deduplicate
  if (raw.task_results && Array.isArray(raw.task_results)) {
    const processedResults: any[] = [];
    const seenKeys = new Set<string>();
    
    for (const tr of raw.task_results) {
      if (tr && typeof tr === 'object') {
        const key = tr.key || '';
        // Skip duplicates and empty keys
        if (key && !seenKeys.has(key)) {
          seenKeys.add(key);
          processedResults.push({
            key: key,
            value: tr.value || ''
          });
        }
      }
    }
    raw.task_results = processedResults;
  }

  return ChatGuideReplySchema.parse(raw);
}

